### ISSUE TITLE
*: avoid long non-preemptable function calls

### ISSUE THREAD

**Issue by dt** [cockroachdb/cockroach#115192]
It appears that due to https://github.com/golang/go/issues/64417, we sometimes see long gc pause times and traces show STW pauses overlapping with block hashing. This has been observed for example [during backups to s3](https://github.com/cockroachdb/cockroach/issues/115190), wherein the s3 sdk does both md5 and sha256 hashes of the 8mb chunks it buffered before uploading, but could affect other users of the hashing libraries as well.

We may want to consider patching the hashing functions on our go fork until the issue is resolved upstream.

Jira issue: CRDB-33925

**Comment by sumeerbhola** [comment]
> wherein the s3 sdk does both md5 and sha256 hashes of the 8mb chunks it buffered before uploading

@dt do you know how long (roughly) the non-preemptable intervals are? < 10ms?

**Comment by dt** [comment]
> how long (roughly) the non-preemptable intervals are? < 10ms?

 typically around or under 10ms but some traces indicated sometimes as much as 40ms

**Comment by nvanbenschoten** [comment]
@dt @sumeerbhola what is our current plan with this? Who is owning its resolution? We have at least one customer who is actively waiting on improvements here, now that we've identified what those improvements should be.

**Comment by sumeerbhola** [comment]
I don't think this should have the AC label, since this is about work that is not preemptible by the goroutine scheduler.

There is a secondary issue that preemptible work should request admission, via the elastic CPU tokens mechanism -- that is already tracked in https://github.com/cockroachdb/cockroach/issues/107770

**Comment by dt** [comment]
> what is our current plan with this

On the DR side, we've determined that we don't have the option to simply avoid hashing -- turning off md5'ing in s3 uploads breaks when bucket versioning locking is enabled -- and the minimum buffer size we can pass to s3 to hash and upload is 5MB (the default is 8 in 23.1, and will be 5MB in 23.2).

> work that is not preemptible by the goroutine scheduler.

I think this issue is tracking making some classes of work that is not preemptable actually be preemptable, i.e. patching our fork of the stdlib so that the two mentioned hashing functions become preemptible (by the scheduler) work. I believe AC is already maintaining our patched fork of go (for the grunning patches?).

>  preemptible work should request admission, via the elastic CPU tokens mechanism

This is on DR's backlog, to integrate AC's existing libraries for requesting tokens/pacing into the the backup process as it constructs SSTs and writes them into the various cloud SDK uploaders, but that won't help us here: we can write as slowly as we want and they uploaders will buffer a chunk before hashing it and uploading it, and the minimum chunk size is 5MB, so that's the unit that'll get passed to hashing at once.

**Comment by nicktrav** [comment]
Shall we add a `T-` label here, so this doesn't fall through the cracks?
